icelake/src/io_v2/functional_writer/precompute_partition_writer.rs

```diff
@@ -0,0 +1,305 @@
+    pub fn try_new(inner: B, partition_type: Any, partition_col_idx: usize) -> Result<Self> {
+        if let Any::Struct(_) = partition_type {
+            Ok(Self {
+                inner,
+                partition_type,
+                partition_col_idx,
+            })
+        } else {
+            Err(Error::new(
+                ErrorKind::IcebergDataInvalid,
+                "Expect partition type to be struct",
+            ))
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<B: IcebergWriterBuilder> IcebergWriterBuilder for PrecomputedPartitionedWriterBuilder<B> {
+    type R = PrecomputedPartitionedWriter<B>;
+
+    async fn build(self, schema: &SchemaRef) -> Result<Self::R> {
+        let mut new_schema_builder = SchemaBuilder::from(schema.as_ref().clone());
+        let remove_field = new_schema_builder.remove(self.partition_col_idx);
+
+        // Check the remove type is the same as the partition type.
+        let expect_partition_type: DataType = self
+            .partition_type
+            .clone()
+            .try_into()
+            .map_err(|e| Error::new(ErrorKind::ArrowError, format!("{}", e)))?;
+        let expected_fields = if let DataType::Struct(fields) = &expect_partition_type {
+            fields
+        } else {
+            unreachable!()
+        };
+        let is_same = if let DataType::Struct(fields) = remove_field.data_type() {
+            fields
+                .iter()
+                .zip_eq(expected_fields.iter())
+                .all(|(f1, f2)| f1.data_type() == f2.data_type())
+        } else {
+            return Err(Error::new(
+                ErrorKind::IcebergDataInvalid,
+                "Expect partition type to be struct",
+            ));
+        };
+        if !is_same {
+            return Err(Error::new(
+                ErrorKind::ArrowError,
+                format!(
+                    "Expect partition type {:?}, but got {:?}",
+                    expect_partition_type,
+                    remove_field.data_type()
+                ),
+            ));
+        }
+
+        let arrow_partition_type_fields =
+            if let DataType::Struct(fields) = self.partition_type.clone().try_into()? {
+                fields
+            } else {
+                unreachable!()
+            };
+        let row_converter = RowConverter::new(vec![SortField::new(DataType::Struct(
+            arrow_partition_type_fields.clone(),
+        ))])
+        .map_err(|e| Error::new(ErrorKind::ArrowError, e.to_string()))?;
+
+        Ok(PrecomputedPartitionedWriter {
+            inner_writers: HashMap::new(),
+            inner_buidler: self.inner,
+            schema: new_schema_builder.finish().into(),
+            partition_type: self.partition_type,
+            row_converter,
+            partition_col_idx: self.partition_col_idx,
+        })
+    }
+}
+
+/// Partition append only writer
+pub struct PrecomputedPartitionedWriter<B: IcebergWriterBuilder> {
```

Comment:
Please add comments here to explain the motivation behind this writer, and what's more important, the expected layout of inputs.